examples/notification-demo/echo_server/echo.c

```diff
@@ -0,0 +1,1975 @@
+
+    if (ret < 0)
+        exit(1);
+
+    switch (pton_ret) {
+    case 1:
+        break;
+    case 0:
+        exit(1);
+    case -1:
+    default:
+        exit(1);
+    }
+
+    if (signal(SIGPIPE, SIG_IGN) == SIG_ERR) {
+        perror("signal(SIGPIPE, SIG_IGN)");
+        exit(1);
+    }
+
+    inbound_sa.sin_family = AF_INET;
+    inbound_sa.sin_port = htons(admin_listen_port);
+
+    inbound_fd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+    if (inbound_fd < 0) {
+        perror("socket(AF_INET, SOCK_STREAM, IPPROTO_TCP");
+        exit(1);
+    }
+
+    {
+        int optval = 1;
+        if (setsockopt(inbound_fd, SOL_SOCKET, SO_REUSEADDR, (void *)&optval,(int)sizeof optval)<0) {
+            perror("setsockopt(..., SO_REUSEADDR, ...)");
+            exit(1);
+        }
+    }
+
+    if (bind(inbound_fd, (const struct sockaddr *)&inbound_sa, sizeof inbound_sa) < 0) {
+        perror("bind");
+        exit(1);
+    }
+
+    if (listen(inbound_fd, 5) < 0) {
+        perror("listen");
+        exit(1);
+    }
+
+    return 0;
+}
+
+static int http_write_header(WOLFSSL *ssl, char *http_buf, size_t http_buf_size, int http_code, ssize_t content_length, const char *content_type, const char *body) {
+    time_t now = time(NULL);
+    struct tm now_tm;
+    char now_formatted[32];
+    size_t now_formatted_len;
+    int i;
+    const char *http_code_description = NULL, *http_code_explanation = NULL;
+    int resp_len, ret;
+
+    static const struct { int code; char *description; char *explanation; } http_codes[] = {
+        { 200, "OK", "Operation succeeded\n" },
+        { 403, "Forbidden", "Authorization denied.\n" },
+        { 404, "Not Found", "Requested path not recognized.\n" },
+        { 500, "Internal Server Error", "Internal server error.\n" }
+    };
+
+    for (i = 0; i < (int)(sizeof http_codes / sizeof http_codes[0]); ++i) {
+        if (http_codes[i].code == http_code) {
+            http_code_description = http_codes[i].description;
+            http_code_explanation = http_codes[i].explanation;
+            break;
+        }
+    }
+    if (http_code_description == NULL)
+        http_code_description = "Unknown code";
+
+    if (gmtime_r(&now, &now_tm) == NULL) {
+        fprintf(stderr, "gmtime_r failed\n");
+        return -1;
+    }
+
+    now_formatted_len = strftime(now_formatted, sizeof now_formatted, "%a, %d %h %Y %H:%M:%S %Z", &now_tm);
+    if (now_formatted_len == 0) {
+        fprintf(stderr, "strftime failed\n");
+        return -1;
+    }
+
+    if ((content_length < 0) && body)
+        content_length = strlen(body);
+
+    if ((content_length == 0) && http_code_explanation) {
+        content_type = "text/plain";
+        content_length = strlen(http_code_explanation);
+        body = http_code_explanation;
+    } else if (! body)
+        body = "";
+
+    resp_len = snprintf(http_buf, http_buf_size, "HTTP/1.0 %d %s\r\n" \
+        "Date: %.*s\r\nServer: Helicopter\r\nContent-Length: %zu%s%s\r\n\r\n%s",
```

Comment:
need to change `Helicopter` to something like `LogServer` -- in fact it occurs to me this file should be named log_server.c, and the parent directory should be log_server/.

Reply:
Renamed